File: ejemplo-graphql/frontend/src/types/Producto.ts
/**
 * Tipos TypeScript para la entidad Producto.
 *
 * Estos tipos reflejan el schema GraphQL del backend.
 * TypeScript nos da seguridad de tipos en tiempo de compilacion,
 * evitando errores comunes como acceder a campos que no existen.
 */

/** Producto tal como llega desde la API GraphQL */
export interface Producto {
  id: string;
  nombre: string;
  descripcion?: string; // Opcional (puede ser null en GraphQL)
  precio: number;
  stock: number;
  categoria?: string; // Opcional
}

/** Datos necesarios para crear un producto (sin ID, lo genera MongoDB) */
export interface ProductoInput {
  nombre: string;
  descripcion?: string;
  precio: number;
  stock: number;
  categoria?: string;
}

File: ejemplo-graphql/src/main/java/com/eia/graphql/model/Producto.java
package com.eia.graphql.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

/**
 * Entidad Producto que representa un producto en la tienda.
 * 
 * Se almacena en la coleccion "productos" de MongoDB.
 * Utiliza anotaciones de Lombok para reducir codigo repetitivo:
 * - @Data genera getters, setters, toString, equals y hashCode
 * - @Builder permite construir objetos con el patron Builder
 * - @NoArgsConstructor genera un constructor sin argumentos
 * - @AllArgsConstructor genera un constructor con todos los campos
 * 
 * @author Universidad EIA - Ingenieria de Sistemas
 */
@Document(collection = "productos") // Mapea esta clase a la coleccion "productos" en MongoDB
@Data                               // Lombok: genera getters, setters, toString, equals, hashCode
@Builder                            // Lombok: habilita el patron Builder para crear instancias
@NoArgsConstructor                  // Lombok: genera constructor vacio (requerido por Spring Data)
@AllArgsConstructor                 // Lombok: genera constructor con todos los campos
public class Producto {

    /**
     * Identificador unico del producto.
     * MongoDB lo genera automaticamente como un ObjectId.
     */
    @Id
    private String id;

    /**
     * Nombre del producto.
     * Campo obligatorio en el schema de GraphQL.
     */
    private String nombre;

    /**
     * Descripcion detallada del producto.
     * Campo opcional en el schema de GraphQL.
     */
    private String descripcion;

    /**
     * Precio del producto en pesos colombianos.
     * Campo obligatorio en el schema de GraphQL.
     */
    private Double precio;

    /**
     * Cantidad de unidades disponibles en inventario.
     * Campo obligatorio en el schema de GraphQL.
     */
    private Integer stock;

    /**
     * Categoria a la que pertenece el producto.
     * Ejemplos: "Electronica", "Ropa", "Alimentos".
     * Campo opcional en el schema de GraphQL.
     */
    private String categoria;
}
